File: backend/app/phone.py
"""Phone number normalization and the country allow-list.

Single source of truth for the Saudi-only launch constraint. The allow-list is
DATA (a list of prefixes + per-prefix format), so extending to other GCC
countries is a config change here — never a Saudi condition scattered across the
codebase. The server guard (SMS hook) and the DB CHECK both derive from this;
the Flutter client mirrors the same rules.
"""
import re

# Extend this list to expand coverage (e.g. add '+9715' for UAE mobiles).
# Each entry: (E.164 prefix, full-format regex for that prefix).
_ALLOW: tuple[tuple[str, "re.Pattern[str]"], ...] = (
    ("+9665", re.compile(r"^\+9665\d{8}$")),  # Saudi mobile: +966 5X XXX XXXX
)

ALLOWED_PHONE_PREFIXES: tuple[str, ...] = tuple(p for p, _ in _ALLOW)


def normalize_phone(raw: str | None) -> str | None:
    """Best-effort E.164 normalization.

    Local Saudi forms (``05XXXXXXXX``, ``5XXXXXXXX``) are interpreted as +966.
    Returns None when the input cannot form a plausible E.164 number. Allow-list
    enforcement is separate — see :func:`is_allowed_phone`.
    """
    if not raw:
        return None
    s = re.sub(r"[^\d+]", "", raw)
    if not s:
        return None

    if s.startswith("00"):
        e164 = "+" + s[2:]
    elif s.startswith("+"):
        e164 = s
    elif s.startswith("966"):
        e164 = "+" + s
    elif s.startswith("0"):
        # local trunk form, e.g. 05XXXXXXXX -> national 5XXXXXXXX under +966
        e164 = "+966" + s[1:]
    elif s.startswith("5") and len(s) == 9:
        # bare Saudi national mobile
        e164 = "+966" + s
    else:
        e164 = "+" + s

    if not re.fullmatch(r"\+\d{8,15}", e164):
        return None
    return e164


def is_allowed_phone(e164: str | None) -> bool:
    """True only for numbers on the allow-list AND matching that country's
    full format (prefix alone is not enough — a wrong-length +9665 fails)."""
    if not e164:
        return False
    return any(e164.startswith(prefix) and rx.match(e164) for prefix, rx in _ALLOW)

File: frontend/lib/core/logic/phone.dart
/// Client layer of the Saudi-only phone guard (mirrors backend `app/phone.py`).
/// Cosmetic — the server SMS hook and the DB CHECK are the real walls — but a
/// Saudi user typing `05..` or `5..` must be accepted and normalized, and a
/// non-Saudi number rejected with a clear message.
///
/// Keep [kAllowedPhonePrefixes] in sync with the backend allow-list when GCC
/// expansion happens (one line, both sides).
const List<String> kAllowedPhonePrefixes = ['+9665'];

final RegExp _saudiMobile = RegExp(r'^\+9665\d{8}$');
final RegExp _e164 = RegExp(r'^\+\d{8,15}$');

/// Best-effort E.164 normalization; local Saudi forms are interpreted as +966.
String? normalizeSaudiPhone(String? raw) {
  if (raw == null) return null;
  var s = raw.replaceAll(RegExp(r'[^\d+]'), '');
  if (s.isEmpty) return null;

  String e164;
  if (s.startsWith('00')) {
    e164 = '+${s.substring(2)}';
  } else if (s.startsWith('+')) {
    e164 = s;
  } else if (s.startsWith('966')) {
    e164 = '+$s';
  } else if (s.startsWith('0')) {
    e164 = '+966${s.substring(1)}';
  } else if (s.startsWith('5') && s.length == 9) {
    e164 = '+966$s';
  } else {
    e164 = '+$s';
  }
  return _e164.hasMatch(e164) ? e164 : null;
}

bool isAllowedSaudiPhone(String? e164) => e164 != null && _saudiMobile.hasMatch(e164);
